wnfs/src/private/hamt/merge.rs

```diff
@@ -0,0 +1,214 @@
+use super::{diff, ChangeType, Node};
+use crate::{BlockStore, FsError, Hasher, Link};
+use anyhow::Result;
+use serde::de::DeserializeOwned;
+use std::{hash::Hash, rc::Rc};
+
+//--------------------------------------------------------------------------------------------------
+// Functions
+//--------------------------------------------------------------------------------------------------
+
+/// Merges a node with another with the help of a resolver function.
+pub async fn merge<K, V, H, F, B: BlockStore>(
+    main_link: Link<Rc<Node<K, V, H>>>,
+    other_link: Link<Rc<Node<K, V, H>>>,
+    f: F,
+    store: &mut B,
+) -> Result<Rc<Node<K, V, H>>>
+where
+    F: Fn(&V, &V) -> Result<V>,
+    K: DeserializeOwned + Eq + Clone + Hash + AsRef<[u8]>,
+    V: DeserializeOwned + Eq + Clone,
+    H: Hasher + Clone + 'static,
+{
+    let kv_changes = diff::kv_diff(main_link.clone(), other_link.clone(), None, store).await?;
+
+    let main_node = main_link.resolve_owned_value(store).await?;
+    let other_node = other_link.resolve_owned_value(store).await?;
+
+    let mut merge_node = Rc::clone(&main_node);
+    for change in kv_changes {
+        match change.r#type {
+            ChangeType::Remove => {
+                merge_node = merge_node
+                    .set(change.key, change.other_value.unwrap(), store)
+                    .await?;
+            }
+            ChangeType::Modify => {
+                let main_value = main_node
+                    .get(&change.key, store)
+                    .await?
+                    .ok_or(FsError::KeyNotFoundInHamt)?;
+
+                let other_value = other_node
+                    .get(&change.key, store)
+                    .await?
+                    .ok_or(FsError::KeyNotFoundInHamt)?;
+
+                merge_node = merge_node
+                    .set(change.key, f(main_value, other_value)?, store)
+                    .await?;
+            }
+            _ => (),
+        }
```

Comment:
This works, but I think it can be more efficient:

When you generate the `kv_diff`, you've walked the tree from the root down a bunch of times.

When you `main_node.get` and `other_node.get`, you walk down from the root given a key again. (you also have to deal with these keys potentially missing, even though you *know* they *should* be there.

In my mind, I'd have written `merge` as yet another recursive function, although I see the value in trying to re-use existing things. (This makes me think: Is there perhaps another abstraction that unifies diffing and merging?)
That function would merge nodes by bitor-ing the `bitfield`, and merging node "pointers" together where both bitfields share a one. That merging would try to merge two `Pointers::Values` by just concatenating & sorting them, and seeing if they need to be broken out into their own node, and it would merge `Pointers::Values` with `Pointers::Link` by turning the `Values` into a new node and running `merge` for nodes recursively (similarily for two `Pointers::Link`s).

Anyway. We can keep the theoretical performance improvement in mind until we *actually* need it.
There's definitely tangible benefit in instead having some code reuse. :+1:

Reply:
```This makes me think: Is there perhaps another abstraction that unifies diffing and merging?)```

I had this question throughout this PR as well, as it seems like two separate facilities that are less linked than they should be? But, yeah, further improvements.

Reply:
Already working on this but I don't want it to hold this PR since it is a performance upgrade thing and does not change public API. I have created a new [issue](https://github.com/wnfs-wg/rs-wnfs/issues/127) for it.